pkg/github/token.go

```diff
@@ -70,59 +78,104 @@ func ParseGitHubTokens() error {
 			tokenKey := splitTokenKeyValuePair[0]
 			tokenValue := splitTokenKeyValuePair[1]
 
-			if Tokens[tokenKey] != "" {
+			if Clients[tokenKey] != nil {
 				return fmt.Errorf("a token with the key '%s' already exists. Each token must have a unique key", tokenKey)
 			}
-			Tokens[tokenKey] = tokenValue
+
+			ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: tokenValue})
+			tc := oauth2.NewClient(context.Background(), ts)
+			token, err := createGitHubClientFromToken(&tc.Transport, tokenValue, tokenKey)
+			if err != nil {
+				return err
+			}
+			Clients[tokenKey] = token
 		}
 	}
 
 	return nil
 }
 
-// getRandomToken randomly retrieves a token from all of the tokens available to HAS
+// getRandomClient randomly retrieves a token from all of the tokens available to HAS
 // It returns the token and the name/key of the token
-func getRandomToken() (string, string, error) {
-	if len(Tokens) == 0 {
-		return "", "", fmt.Errorf("no GitHub tokens initialized")
+func getRandomClient(clientPool map[string]*GitHubClient) (*GitHubClient, error) {
+	if len(clientPool) == 0 {
```

Comment:
Will we ever get primary rate limited if we're doing preemptive checks?  If not, then I'll have to change the way the metrics work

Reply:
This will work for secondary rate limiting too.  Can we pass in a couple of booleans to indicate if we've exhausted the tokens for primary or secondary rate limiting?

Reply:
No, the Rate Limit API endpoint is not rate limited.

> This will work for secondary rate limiting too. Can we pass in a couple of booleans to indicate if we've exhausted the tokens for primary or secondary rate limiting?

I'm not sure what you mean here? What would the booleans be used for?

Reply:
That we exhausted the tokens for primary rate limit or secondary rate limits.  I can alternatively do a check on the length of the clientPool before the recursive call but seeing how there's a check done at the beginning, I thought we could avoid duplication.

Reply:
Right okay. So it's for metrics? What if we just add the necessary lines when we detect the primary/secondary rate limiting, instead of using bools to track?

Reply:
yeah, that's fine.  I can workaround it.  I was planning to add metrics if we are close to exhausting (~2 tokens left) and when we have no tokens left so we can get the alerts